blaze-server/src/main/scala/org/http4s/blaze/server/Http1ServerStage.scala

```diff
@@ -303,17 +302,21 @@ private[blaze] class Http1ServerStage[F[_]](
             logger.trace("Request/route requested closing connection.")
             F.delay(closeConnection())
           } else
-            F.delay {
-              bodyCleanup().onComplete {
-                case s @ Success(_) => // Serve another request
-                  parser.reset()
-                  handleReqRead(s)
-
-                case Failure(EOF) => closeConnection()
-
-                case Failure(t) => fatalError(t, "Failure in body cleanup")
-              }(trampoline)
-            }
+            F.async_[Unit] { cb =>
+              cb(
+                Right(
+                  bodyCleanup().onComplete {
```

Comment:
Hmm. Most likely you're right. Though classically applying `async` (with the handling of `onComplete` branches) doesn't fit well because we have extra processing. I will think a little bit more about this one.

Reply:
What do you think about this callback then:
```scala
F.async_[Unit] { cb =>
  bodyCleanup().onComplete {
    case s @ Success(_) => // Serve another request
      cb(Right {
        parser.reset()
        handleReqRead(s)
      })

    case Failure(EOF) =>
      cb(Right(closeConnection()))

     case Failure(t) =>
       cb(Right(fatalError(t, "Failure in body cleanup")))
    }(trampoline)
}
```
I think that failure cases of the `bodyCleanup().onComplete` should go to success cases of callback. Because we don't want to handle errors of newly created effect.

Reply:
Actually, I don't like this part. That's what I meant above.
```scala
cb(Right {
  parser.reset()
  handleReqRead(s)
})
```

Reply:
Hmm. Maybe something like this?

```scala
F.async_[Unit] { cb =>
  case Success(_) => cb(Right(...))
  case Failure(t) => cb(Left(...))
}
.attempt
.handleErrorWith {
 ...
}

Reply:
That could also mean perf reduce because of async-boundaries. Isn't it?

Reply:
Firstly, I edited it because I wrote it wrong 😆 but, I think you got the idea.

> That could also mean perf reduce because of async-boundaries.

I can't say for certain, because IDK anything about blaze. But, as you said yourself, a good strategy is always to get onto the CE compute pool ASAP because it is much more performant.

Reply:
After some suss out, I think *that* async-boundary could break some semantics of blaze request processing. Performance for real could stay the same (or, maybe, have some benefits, who know). So I vote for my proposed variant. It would be cool to get more opinions on this.